tests/typehints/test_script_decorator.py

```diff
@@ -0,0 +1,109 @@
+from pathlib import Path
+from subprocess import run
+from tempfile import TemporaryDirectory
+from textwrap import dedent
+from typing import Iterator, Tuple
+
+import pytest
+
+from hera.workflows import Step, Task
+
+SIMPLE_SCRIPT = """
+from hera.workflows import script
+
+@script()
+def simple_script(input: str) -> int:
+    print(input)
+    return len(input)
+""".strip()
+
+
+def run_mypy(python_code: str):
+    with TemporaryDirectory() as d:
+        python_file = Path(d) / "example.py"
+        python_file.write_text(python_code)
+        mypy_cmd = ["mypy", "--config-file", "tests/typehints/test-mypy.toml", str(python_file)]
+        result = run(mypy_cmd, check=False, capture_output=True, encoding="utf-8")
+        if result.returncode != 0:
+            msg = f"Error calling {' '.join(mypy_cmd)}:\n{result.stderr}{result.stdout}"
+            raise AssertionError(msg)
+        return result.stdout.replace(d, "")
+
+
+def test_underlying_function_args_invocation():
+    """Verify the underlying implementation of a script can be invoked with positional arguments."""
+    STEP_INVOCATION = """
+        result = simple_script("Hello World!")
+        reveal_type(result)
+    """
+    result = run_mypy(SIMPLE_SCRIPT + dedent(STEP_INVOCATION))
+    assert 'Revealed type is "builtins.int"' in result
+
+
+def test_underlying_function_kwargs_invocation():
+    """Verify the underlying implementation of a script can be invoked with named arguments."""
+    STEP_INVOCATION = """
+        result = simple_script(input="Hello World!")
+        reveal_type(result)
+    """
+    result = run_mypy(SIMPLE_SCRIPT + dedent(STEP_INVOCATION))
+    assert 'Revealed type is "builtins.int"' in result
+
+
+def test_invocation_with_no_arguments():
+    """Verify a script can be invoked with no arguments.
+
+    Without knowing the invocation context, which the type system does not have access to,
+    the return type could be None, Step or Task.
+    """
+    STEP_INVOCATION = """
+        result = simple_script()
+        reveal_type(result)
+    """
+    result = run_mypy(SIMPLE_SCRIPT + dedent(STEP_INVOCATION))
+    assert 'Revealed type is "Union[hera.workflows.steps.Step, hera.workflows.task.Task, None]"' in result
+
+
+def test_parameter_named_name():
+    """If a script has a 'name' argument, the Step/Task overload will take precedence."""
```

Comment:
I don't think this is the right way - if we take a scenario where a user is adding the Hera library to their existing code and adds the `@script` decorator to a function, the user's lint checks will break. We are an optional wrapper and should be treating the code as such.